src/tls.jl

```diff
@@ -0,0 +1,100 @@
+struct TaskLocalState
+    device::ROCDevice
+    queue::ROCQueue
+    priority::Symbol
+    function TaskLocalState(device::Union{ROCDevice,Nothing},
+                            queue::Union{ROCQueue,Nothing},
+                            priority::Symbol)
+        if device === nothing
+            if queue === nothing
+                device = Runtime.get_default_device()
+                queue = ROCQueue(device; priority, pooled=true)
+            else
+                device = AMDGPU.device(queue)
+            end
+        else
+            if queue === nothing
+                queue = ROCQueue(device; priority, pooled=true)
+            else
+                @assert queue.priority == priority
+            end
+        end
+        return new(device, queue, priority)
+    end
+end
+TaskLocalState() = TaskLocalState(nothing, nothing, :normal)
+
+function Base.show(io::IO, state::TaskLocalState)
+    println(io, "TaskLocalState:")
+    println(io, "  Device: $(state.device)")
+    println(io, "  Queue: $(state.queue)")
+    print(io, "  Priority: $(state.priority)")
+end
+
+"""
+    task_local_state() -> TaskLocalState
+
+Returns the task-local state in the form of a `TaskLocalState`. Automatically
+picks a device and queue if they haven't already been selected.
+"""
+task_local_state() =
+    get!(()->TaskLocalState(), task_local_storage(), :__amdgpu_task_state)
```

Comment:
```suggestion
    get!(()->TaskLocalState(), task_local_storage(), :__amdgpu_task_state)::TaskLocalState
```

I think this is necessary?